File: src/gateway/runtime_inbox.py
"""Bounded, read-only projection for a Desktop cross-runtime inbox.

Each Gateway remains authoritative for its own sessions and durable ledgers.
Desktop may aggregate several of these snapshots, but it must not rediscover
status rules or copy unbounded task/goal payloads across projects.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Iterable

_SESSION_STATUSES = {"needs_input", "failed", "working", "queued", "idle", "inactive", "completed"}
_GOAL_STATUSES = {"draft", "active", "blocked", "failed", "achieved"}
_TASK_STATUSES = {"queued", "running", "cancelling", "failed", "paused", "interrupted", "done", "cancelled"}
_DECISION_KINDS = {"confirmation", "goal", "task", "run", "pr_check", "pr_review", "hook"}
_DECISION_SEVERITIES = {"critical", "high", "medium", "low"}
_DECISION_ACTIONS = {"confirm", "open_goal", "resume_task", "open_task", "open_run", "open_diff", "open_session"}


def _get(item: Any, name: str, default: Any = "") -> Any:
    if isinstance(item, dict):
        return item.get(name, default)
    return getattr(item, name, default)


def _text(value: Any, limit: int) -> str:
    value = str(value or "").strip()
    return value if len(value) <= limit else value[:limit] + "…"


def _session_view(item: Any) -> dict[str, Any] | None:
    sid = _text(_get(item, "sid"), 120)
    if not sid:
        return None
    status = _text(_get(item, "status"), 24)
    if status not in _SESSION_STATUSES:
        status = "inactive"
    background = _get(item, "background_tasks", {})
    background = background if isinstance(background, dict) else {}
    hook_issues = _get(item, "hook_issues", {})
    hook_issues = hook_issues if isinstance(hook_issues, dict) else {}
    context = _get(item, "context", {})
    context = context if isinstance(context, dict) else {}
    return {
        "sid": sid,
        "title": _text(_get(item, "title") or "新任务", 120),
        "status": status,
        "updated": max(0.0, float(_get(item, "updated", 0) or 0)),
        "pending_input_count": max(0, int(_get(item, "pending_input_count", 0) or 0)),
        "queue_count": max(0, int(_get(item, "queue_count", 0) or 0)),
        "activity": _text(_get(item, "activity"), 160),
        "running_prompt": _text(_get(item, "running_prompt"), 160),
        "cwd": _text(_get(item, "cwd"), 500),
        "branch": _text(_get(item, "branch"), 160),
        "background_tasks": {
            "active": max(0, int(background.get("active") or 0)),
            "attention": max(0, int(background.get("attention") or 0)),
            "latest_task_id": _text(background.get("latest_task_id"), 180),
            "active_task_id": _text(background.get("active_task_id"), 180),
            "attention_task_id": _text(background.get("attention_task_id"), 180),
        },
        "hook_issues": {"count": max(0, int(hook_issues.get("count") or 0))},
        "context": {
            "pct": max(0, min(999, int(context.get("pct") or 0))),
            "used_tokens": max(0, int(context.get("used_tokens") or 0)),
            "max_tokens": max(0, int(context.get("max_tokens") or 0)),
        },
    }


def _goal_view(item: Any) -> dict[str, Any] | None:
    goal_id = _text(_get(item, "id"), 180)
    status = _text(_get(item, "status"), 24)
    if not goal_id or status not in _GOAL_STATUSES:
        return None
    criteria = _get(item, "acceptance_criteria", [])
    criteria = criteria if isinstance(criteria, list) else []
    passed = sum(_get(criterion, "status") == "passed" for criterion in criteria)
    failed = sum(_get(criterion, "status") == "failed" for criterion in criteria)
    return {
        "id": goal_id,
        "objective": _text(_get(item, "objective"), 240),
        "status": status,
        "blocker": _text(_get(item, "blocker"), 500),
        "next_action": _text(_get(item, "next_action"), 300),
        "updated": _text(_get(item, "updated") or _get(item, "created"), 80),
        "progress": {"passed": passed, "failed": failed, "total": len(criteria)},
    }


def _decision_view(item: Any) -> dict[str, Any] | None:
    decision_id = _text(_get(item, "id"), 240)
    kind = _text(_get(item, "kind"), 24)
    if not decision_id or kind not in _DECISION_KINDS:
        return None
    severity = _text(_get(item, "severity"), 24)
    if severity not in _DECISION_SEVERITIES:
        severity = "high"
    action = _text(_get(item, "action"), 32)
    if action not in _DECISION_ACTIONS:
        action = "confirm" if kind == "confirmation" else "open_session"
    return {
        "id": decision_id,
        "kind": kind,
        "severity": severity,
        "title": _text(_get(item, "title") or "需要处理", 200),
        "detail": _text(_get(item, "detail"), 1200),
        "created": _text(_get(item, "created"), 80),
        "target_id": _text(_get(item, "target_id") or decision_id, 240),
        "action": action,
        "session_id": _text(_get(item, "session_id"), 180),
        "tainted": bool(_get(item, "tainted", False)),
        "can_dismiss": bool(_get(item, "can_dismiss", False)),
    }


def _task_view(item: Any) -> dict[str, Any] | None:
    task_id = _text(_get(item, "id"), 180)
    status = _text(_get(item, "status"), 24)
    if not task_id or status not in _TASK_STATUSES:
        return None
    return {
        "id": task_id,
        "status": status,
        "prompt": _text(_get(item, "prompt"), 240),
        "detail": _text(_get(item, "error") or _get(item, "result"), 500),
        "owner_session": _text(_get(item, "owner_session"), 180),
        "goal_id": _text(_get(item, "goal_id"), 180),
        "plan_id": _text(_get(item, "plan_id"), 180),
        "branch": _text(_get(item, "branch"), 180),
        "updated": _text(_get(item, "updated") or _get(item, "created"), 80),
    }


def build_runtime_inbox(
    *,
    scope: str,
    sessions: Iterable[Any] = (),
    decisions: Iterable[Any] = (),
    goals: Iterable[Any] = (),
    tasks: Iterable[Any] = (),
) -> dict[str, Any]:
    """Return a stable summary suitable for polling across local runtimes."""
    session_views = [view for item in list(sessions)[:100] if (view := _session_view(item))]
    decision_views = [view for item in list(decisions)[:100] if (view := _decision_view(item))]
    goal_views = [view for item in list(goals)[:200] if (view := _goal_view(item))]
    task_views = [view for item in list(tasks)[:200] if (view := _task_view(item))]
    return {
        "version": 1,
        "scope": _text(scope, 24),
        "generated_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "sessions": session_views,
        "decisions": decision_views,
        "goals": goal_views,
        "tasks": task_views,
        "counts": {
            "sessions_needing_input": sum(item["status"] == "needs_input" for item in session_views),
            "sessions_working": sum(item["status"] == "working" for item in session_views),
            "sessions_queued": sum(item["status"] == "queued" for item in session_views),
            "decisions": len(decision_views),
            "hook_issues": sum(item["hook_issues"]["count"] for item in session_views),
            "goals_active": sum(item["status"] in {"draft", "active"} for item in goal_views),
            "goals_blocked": sum(item["status"] in {"blocked", "failed"} for item in goal_views),
            "tasks_active": sum(item["status"] in {"queued", "running", "cancelling"} for item in task_views),
            "tasks_attention": sum(item["status"] in {"failed", "paused", "interrupted"} for item in task_views),
        },
    }

File: desktop/src/lib/runtimeInbox.ts
/**
 * 跨项目收件箱的状态归类——把"还没起来"和"连上过又断了"分开。
 *
 * 真机诊断（2026-09-17）：刚加完项目、runtime 正在启动的那十几秒里，轮询自然失败，界面却直接
 * 报「1 个失联」。失联是要人去查的故障，启动中是等一会儿就好——两者长一个样，用户学会的是
 * 忽略这个提示，于是真失联时也不会去看。
 *
 * 判据只有两条，都不依赖文案匹配：这个 runtime **成功连上过吗**，以及**是不是刚出现**。
 */

export type RuntimeReachability = "ok" | "starting" | "unreachable";

/** 首次出现后多久仍连不上就按失联算。真机上项目 runtime 冷启动约 3 秒，留足余量。 */
export const STARTING_GRACE_MS = 20_000;

export type ReachabilityInput = {
  error?: string;
  firstSeenAt: number;
  everReachable: boolean;
};

export function classifyReachability(item: ReachabilityInput, now: number): RuntimeReachability {
  if (!item.error) return "ok";
  if (item.everReachable) return "unreachable";   // 连上过 → 现在连不上就是真故障
  return now - item.firstSeenAt < STARTING_GRACE_MS ? "starting" : "unreachable";
}

export type InboxCounts = {
  decisions: number;
  hook_issues: number;
  goals_blocked: number;
  goals_active: number;
  tasks_attention: number;
  tasks_active: number;
  sessions_working: number;
  sessions_queued: number;
};

export type SummaryInput = ReachabilityInput & { counts?: InboxCounts | null };

export type InboxSummary = {
  runtimes: number;
  unreachable: number;
  starting: number;
  actionable: number;
  working: number;
  queued: number;
  goals: number;
  tasks: number;
};

export function summarizeRuntimeInboxes(items: SummaryInput[], now: number): InboxSummary {
  const summary: InboxSummary = {
    runtimes: items.length,
    unreachable: 0,
    starting: 0,
    actionable: 0,
    working: 0,
    queued: 0,
    goals: 0,
    tasks: 0,
  };
  items.forEach((item) => {
    const reachability = classifyReachability(item, now);
    if (reachability === "unreachable") summary.unreachable += 1;
    if (reachability === "starting") summary.starting += 1;
    const counts = item.counts;
    if (!counts) return;
    summary.actionable += counts.decisions + counts.hook_issues
      + counts.goals_blocked + counts.tasks_attention;
    summary.working += counts.sessions_working;
    summary.queued += counts.sessions_queued;
    summary.goals += counts.goals_active + counts.goals_blocked;
    summary.tasks += counts.tasks_active + counts.tasks_attention;
  });
  return summary;
}

/** 收件箱副标题那一行：优先说要处理的，其次是真故障，再次是启动中，最后才是执行中。 */
export function runtimeInboxSubtitle(summary: InboxSummary): string {
  const workspaces = `${summary.runtimes} 个工作区`;
  if (summary.actionable > 0) return `${workspaces} · ${summary.actionable} 项待处理`;
  if (summary.unreachable > 0) return `${workspaces} · ${summary.unreachable} 个失联`;
  if (summary.starting > 0) return `${workspaces} · ${summary.starting} 个启动中`;
  return `${workspaces} · ${summary.working} 个执行中`;
}
